### PR TITLE
[FLINK-26813[SQL-API] Supports ADD/MODIFY column/watermark/constraint syntax parse for ALTER TABLE

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*Support ADD column/constraint/watermark for ALTER TABLE statement parse*

## Brief change log

  - *Support ADD column/constraint/watermark for ALTER TABLE statement parse*

## Verifying this change

This change added tests and can be verified as follows:

 - *Added uint tests  in FlinkSqlParserImplTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: ( no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@wuchong Can you help to review this PR?

**Comment by RocMarshal** [comment]
@lsyldliu Thank you very much for your contribution.

However, I have started working on this ticket **https://github.com/apache/flink/pull/17560.**
If you have any ideas, you can discuss them directly on the corresponding pages ([**FLINK-22315**](https://issues.apache.org/jira/browse/FLINK-22315) page or **https://github.com/apache/flink/pull/17560** page ), which is the significance of code review.

It is recommended that we start the next advance after a supported conclusion. If not, this will make the ticket's asset mechanism meaningless and inconvenient for the normal review and management of the community develop work.
If your opinions were supported finally, I'll merge your code on your PR branch to **https://github.com/apache/flink/pull/17560** . Any  suggestions would be appreciated.

**Comment by lsyldliu** [comment]
> @lsyldliu Thank you very much for your contribution.
> 
> However, I have started working on this ticket [FLINK-22315](https://issues.apache.org/jira/browse/FLINK-22315) & **#17560.** If you have any ideas, you can discuss them directly on the corresponding pages ([**FLINK-22315**](https://issues.apache.org/jira/browse/FLINK-22315) page or **#17560** page ), which is the significance of code review.
> 
> It is recommended that we start the next advance after a supported conclusion. If not, this will make the ticket's asset mechanism meaningless and inconvenient for the normal review and management of the community develop work. If your opinions were supported finally, I'll merge your code on your PR branch to **#17560** . Any suggestions would be appreciated.


> Thanks for your feedback, I'm doing the work related [FLINK-21634](https://issues.apache.org/jira/browse/FLINK-21634). I'm sorry for the confuse you've caused, I will discuss with @wuchong offline.

**Comment by RocMarshal** [comment]
Thanks @lsyldliu @wuchong for the discussion offline and reply.
That sounds good to me. 
IMO, it seems more appropriate that we would convert the new ticket FLINK-26813 as a sub-task of FLINK-21634.

Feel free to go ahead.

**Comment by wuchong** [comment]
> IMO, it seems more appropriate that we would convert the new ticket [FLINK-26813](https://issues.apache.org/jira/browse/FLINK-26813) as a sub-task of [FLINK-21634](https://issues.apache.org/jira/browse/FLINK-21634).

Yes, I think we should do that. cc @lsyldliu

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTable.java [hunk 1]
```diff
@@ -94,4 +97,15 @@ public SqlNodeList getPartitionSpec() {
...
+
+    protected void printIndent(SqlWriter writer) {
+        writer.sep(",", false);
+        writer.newlineAndIndent();
+        writer.print("  ");
+    }
+
+    /** Alter table context. */
+    public static class AlterTableContext extends SqlCreateTable.TableCreationContext {
+        public List<SqlNode> columnPositions = new ArrayList<>();
```
Comment (luoyuxia) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTable.java:109]: Do we need `columnPositions`? Seems it is never used.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            writer.endList(frame);
+        } else {
+            if (addedColumns.getList().size() == 1) {
+                // add single column case
+                addedColumns.getList().get(0).unparse(writer, leftPrec, rightPrec);
+            } else if (watermark != null) {
+                // add watermark case
+                watermark.unparse(writer, leftPrec, rightPrec);
+            }
+            // TODO: add constraint case
```
Comment (luoyuxia) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java:147]: Is it still TODO ?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            List<SqlTableConstraint> constraint) {
+        super(pos, tableName, null);
+        this.withParen = withParen;
+        this.addedColumns = addedColumns;
+        this.watermark = sqlWatermark;
+        this.constraint = constraint;
+    }
+
+    /** Returns the column constraints plus the table constraints. */
+    public List<SqlTableConstraint> getFullConstraints() {
```
Comment (luoyuxia) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java:79]: is the method needed? Seems it's nerver used.

**Review by luoyuxia — COMMENTED** [review]
The ci is failed

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java (hunk 2 above)
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java:147]: Yes, current it is need, but I remove it firstly in this PR

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java (hunk 3 above)
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java:79]: remove it first

**Comment by lsyldliu** [comment]
@luoyuxia Thanks for your review, I have updated the comment. Moreover, the failed ci is caused by e2e test, which is not related to this change.

**Comment by lsyldliu** [comment]
@flinkbot run azure

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 4]
```diff
@@ -313,6 +313,175 @@ void testAlterTable() {
...
+        sql("alter table t1 add watermark for ts as ts - interval '1' second")
+                .ok("ALTER TABLE `T1` ADD WATERMARK FOR `TS` AS (`TS` - INTERVAL '1' SECOND)");
+        sql("alter table default_database.t1 add watermark for ts as ts - interval '1' second")
+                .ok(
+                        "ALTER TABLE `DEFAULT_DATABASE`.`T1` ADD WATERMARK FOR `TS` AS (`TS` - INTERVAL '1' SECOND)");
+        sql("alter table default_catalog.default_database.t1 add watermark for ts as ts - interval '1' second")
+                .ok(
+                        "ALTER TABLE `DEFAULT_CATALOG`.`DEFAULT_DATABASE`.`T1` "
+                                + "ADD WATERMARK FOR `TS` AS (`TS` - INTERVAL '1' SECOND)");
+
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:345]: add case for add single constraint here.

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. I left some comments.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTable.java [hunk 5]
```diff
@@ -94,4 +97,15 @@ public SqlNodeList getPartitionSpec() {
     public LinkedHashMap<String, String> getPartitionKVs() {
         return SqlPartitionUtils.getPartitionKVs(getPartitionSpec());
     }
+
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTable.java:100]: I think this is a technical debt, the refactoring is slightly more work, we should refactor it in another separate issue, what do you think of it?

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 6]
```diff
@@ -313,6 +313,175 @@ void testAlterTable() {
         sql(sql3).ok(expected3);
     }
 
+    @Test
+    void testAlterTableAdd() {
+        // add single column
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:318]: The comments just only for better code readability, the same type syntax should be tested in one method, same as other. If split, the test will be very much and readability is not good.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 7]
```diff
@@ -567,12 +568,60 @@ SqlAlterTable SqlAlterTable() :
...
+            (
+                <COMMA> AlterTableAddOrModify(ctx)
+            )*
+            <RPAREN>
+        )
+        {
+            return new SqlAlterTableAdd(
+                        startPos.plus(getPos()),
+                        tableIdentifier,
+                        withParen,
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl:596]: The `withParen ` is just for SqlNode unparse, I agree with you, this seems that the code will much cleaner.

> withParen should be true when add multiple columns? 
No, user maybe `ADD (a  INT)` with single column.

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 8]
```diff
@@ -313,6 +313,175 @@ void testAlterTable() {
...
+        // add compute column
+        sql("alter table t1 add col_int as col_a - col_b after col_b")
+                .ok("ALTER TABLE `T1` ADD `COL_INT` AS (`COL_A` - `COL_B`) AFTER `COL_B`");
+        // add metadata column
+        sql("alter table t1 add col_int int metadata from 'mk1' virtual comment 'comment_metadata' after col_b")
+                .ok(
+                        "ALTER TABLE `T1` ADD `COL_INT` INTEGER METADATA FROM 'mk1' VIRTUAL "
+                                + "COMMENT 'comment_metadata' AFTER `COL_B`");
+
+        // add watermark
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:335]: Currently, defining multiple watermark is not allowed in watermark parse phase, please see https://github.com/apache/flink/blob/master/flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl#L714

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            writer.newlineAndIndent();
+            writer.endList(frame);
+        } else {
+            if (addedColumns.getList().size() == 1) {
+                // add single column case
+                addedColumns.getList().get(0).unparse(writer, leftPrec, rightPrec);
+            } else if (watermark != null) {
+                // add watermark case
+                watermark.unparse(writer, leftPrec, rightPrec);
+            }
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java:124]: After https://issues.apache.org/jira/browse/FLINK-22315 we should unparse it, Single constraint is parsed into SqlAlterTableAddConstraint currnetly.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/position/SqlColumnPosSpec.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+/** Enumeration of SQL column position specification. */
+public enum SqlColumnPosSpec {
+    FIRST("FIRST"),
+    AFTER("AFTER"),
+    LAST("LAST");
+
+    private final String digest;
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/position/SqlColumnPosSpec.java:30]: Indeed, we don't need it, but what do you think about keeping it consistent with the code style of `SqlConstraintEnforcement` and `SqlUniqueSpec`?

**Comment by lsyldliu** [comment]
@fsk119 Thanks for your review, I've addressed your comments.

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java (hunk 4 above)
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:363]: This is covered by `testAlterTable`

**Review by fsk119 — COMMENTED** [review]
Thanks for your update and comments.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/position/SqlTableColumnPosition.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+        column.unparse(writer, leftPrec, rightPrec);
+        if (isFirstColumn()) {
+            positionSpec.unparse(writer, leftPrec, rightPrec);
+        } else if (isAfterReferencedColumn()) {
+            positionSpec.unparse(writer, leftPrec, rightPrec);
+            referencedColumn.unparse(writer, leftPrec, rightPrec);
+        }
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/position/SqlTableColumnPosition.java:104]: Here can't throw exception, it is default behavior. I've add comment here.

**Comment by lsyldliu** [comment]
@fsk119 Thanks for review, I've addressed your comment.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableSchema.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    public List<SqlTableConstraint> getConstraints() {
+        return constraints;
+    }
+
+    @Override
+    public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+        super.unparse(writer, leftPrec, rightPrec);
+    }
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableSchema.java:80]: If it invokes `super`, I think we can remove this.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableSchema.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Nonnull
+    @Override
+    public List<SqlNode> getOperandList() {
+        return ImmutableNullableList.of(
+                getTableName(),
+                columnList,
+                watermark,
+                new SqlNodeList(constraints, SqlParserPos.ZERO));
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableSchema.java:62]: nit: It's better to align with `SqlCreateTable`.

```
return ImmutableNullableList.of(
                getTableName(),
                columnList,
                new SqlNodeList(constraints, SqlParserPos.ZERO),
                watermark);
```

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableSchema.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    protected final SqlNodeList columnList;
+    @Nullable protected final SqlWatermark watermark;
+    protected final List<SqlTableConstraint> constraints;
+
+    public SqlAlterTableSchema(
+            SqlParserPos pos,
+            SqlIdentifier tableName,
+            SqlNodeList columnList,
+            @Nullable SqlWatermark sqlWatermark,
+            List<SqlTableConstraint> constraints) {
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableSchema.java:48]: nit: would be better with the order of `constraints`, `watermark`.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/position/SqlColumnPosSpec.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+package org.apache.flink.sql.parser.ddl.position;
+
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+/** Enumeration of SQL column position specification. */
+public enum SqlColumnPosSpec {
+    FIRST("FIRST"),
+    AFTER("AFTER"),
+    LAST("LAST");
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/position/SqlColumnPosSpec.java:28]: The issue doesn't mention `LAST` and no test covers this. I just think do we need `last` now?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlTableUtils.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.flink.sql.parser.ddl.SqlWatermark;
+import org.apache.flink.sql.parser.ddl.constraint.SqlTableConstraint;
+
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlWriter;
+
+import java.util.List;
+
+/** Utils methods for table DDLs. */
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlTableUtils.java:30]: Utils to unparse DDLs

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlTableUtils.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+public class SqlTableUtils {
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlTableUtils.java:31]: SqlUnparseUtils?

**Review by fsk119 — COMMENTED** [review]
Thanks for your update. LGTM in general. I left some minor suggestions.

**Comment by lsyldliu** [comment]
@fsk119 I've updated your comment, please see again.

**Comment by lsyldliu** [comment]
@flinkbot run azure

**Review by fsk119 — APPROVED** [review]
Thanks for your update. LGTM.

**Pull request merged**